internal/provisioner/plain/controllers/bundle_controller.go

```diff
@@ -464,53 +476,186 @@ func bundleImagePod(pod *corev1.Pod, source rukpakv1alpha1.ImageSource, unpackIm
+		}
+	}
+
+	var bundleFS fs.FS = &billyFS{wt.Filesystem}
+
+	// Subdirectory
+	if gitsource.Directory != "" {
+		directory := gitsource.Directory
+		if directory[0] == '.' {
+			directory = directory[2:]
+		}
+		sub, err := wt.Filesystem.Chroot(directory)
+		if err != nil {
+			return fmt.Errorf("get subdirectory %q for repository %q: %w", gitsource.Directory, gitsource.Repository, err)
+		}
+		bundleFS = &billyFS{sub}
+	}
+
+	// Extract objects
+	objects, err := getObjects(bundleFS)
+	if err != nil {
+		return updateStatusUnpackFailing(u, fmt.Errorf("get objects from bundle manifests: %w", err))
+	}
+	if len(objects) == 0 {
+		return updateStatusUnpackFailing(u, errors.New("invalid bundle: found zero objects: "+
+			"plain+v0 bundles are required to contain at least one object"))
 	}
-	return pod, nil
+
+	// Store bunle
+	if err := r.Storage.Store(ctx, bundle, bundleFS); err != nil {
+		return updateStatusUnpackFailing(u, fmt.Errorf("persist bundle objects: %w", err))
+	}
+
+	contentURL, err := r.Storage.URLFor(ctx, bundle)
+	if err != nil {
+		return updateStatusUnpackFailing(u, fmt.Errorf("get content URL: %w", err))
+	}
+
+	var bundleImageDigest string
+	u.UpdateStatus(
+		updater.EnsureBundleDigest(bundleImageDigest),
+		updater.EnsureContentURL(contentURL),
+		updater.SetPhase(rukpakv1alpha1.PhaseUnpacked),
+		updater.EnsureCondition(metav1.Condition{
+			Type:   rukpakv1alpha1.TypeUnpacked,
+			Status: metav1.ConditionTrue,
+			Reason: rukpakv1alpha1.ReasonUnpackSuccessful,
+		}),
+	)
+
+	return nil
 }
 
-// addUnpackerInitContainer injects the install-unpacker init container into the given pod.
-// addUnpackerInitContainer assumes the pod has an array of init containers initialized.
-func addUnpackerInitContainer(pod *corev1.Pod, unpackImage string) *corev1.Pod {
-	pod.Spec.InitContainers[0].Name = "install-unpacker"
-	pod.Spec.InitContainers[0].Image = unpackImage
-	pod.Spec.InitContainers[0].ImagePullPolicy = corev1.PullIfNotPresent
-	pod.Spec.InitContainers[0].Command = []string{"cp", "-Rv", "/unpack", "/bin/unpack"}
-	pod.Spec.InitContainers[0].VolumeMounts = []corev1.VolumeMount{{Name: "util", MountPath: "/bin"}}
-	return pod
+// billy.Filesysten -> fs.FS
+var (
+	_ fs.FS         = &billyFS{}
+	_ fs.ReadDirFS  = &billyFS{}
+	_ fs.ReadFileFS = &billyFS{}
+	_ fs.StatFS     = &billyFS{}
+	_ fs.File       = &billyFile{}
+)
+
+type billyFS struct {
+	billy.Filesystem
+}
+
+func (f *billyFS) ReadFile(name string) ([]byte, error) {
+	file, err := f.Filesystem.Open(name)
+	if err != nil {
+		return nil, err
+	}
+	return io.ReadAll(file)
+}
+
+func (f *billyFS) Open(path string) (fs.File, error) {
+	file, err := f.Filesystem.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	fi, err := f.Filesystem.Stat(path)
+	return &billyFile{file, fi, err}, nil
+}
+
+func (f *billyFS) ReadDir(name string) ([]fs.DirEntry, error) {
+	fis, err := f.Filesystem.ReadDir(name)
+	if err != nil {
+		return nil, err
+	}
+	entries := []fs.DirEntry{}
```

Comment:
Nit: (probably from my own code? :sweat_smile:) 
```suggestion
	entries := make([]fs.DirEntry, 0, len(fis))
```